### PR TITLE
feat: Implemented event invoice table for organizers

### LINKED ISSUE #3118: Create a list of invoices in the Billing sub tab "Invoices"
In the billing subtab there should be the following invoices with different areas:
1. Ticket Invoices
* Upcoming Events
* Past Events
2. Organizer Invoices
* Overdue Invoices
* Upcoming Invoices
* Paid Invoices

Please refer to the following screenshot for the expected table view:

Screenshot from 2019-06-11 22-57-43

### PR DESCRIPTION
**Pull request by mrsaicharan1**
Fixes #3118 

#### Short description of what this resolves:
This feature adds the tables required for event invoices which shows the upcoming, paid and due event invoices along with relevant information. 
#### Screenshots
As soon as the server implementation for event invoices are functional, the event invoices will show up here along with their respective invoice links.
Screenshot 2019-07-31 at 3 06 51 PM

#### Changes proposed in this pull request:

- Filter upcoming, paid and due invoices
- Added controller for event-invoices
- Designed template for event invoices

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by kushthedude** [comment]
@mrsaicharan1 Please put it on-hold, I dont think you would like to refactor it twice , Ember-Tables are almost ready

File: tests/unit/controllers/account/billing-info/invoices-test.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { module, test } from 'qunit';
+import { setupTest } from 'ember-qunit';
+
+module('Unit | Controller | account/billing-info/invoices', function(hooks) {
+  setupTest(hooks);
+
+  // Replace this with your real tests.
```
Comment (Anupam-dagar) [inline tests/unit/controllers/account/billing-info/invoices-test.js]: This comment is not really needed

File: tests/unit/controllers/account/billing-info/invoices-test.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { module, test } from 'qunit';
+import { setupTest } from 'ember-qunit';
+
+module('Unit | Controller | account/billing-info/invoices', function(hooks) {
```
Comment (Anupam-dagar) [inline tests/unit/controllers/account/billing-info/invoices-test.js]: this test is failing in travis

File: tests/unit/controllers/account/billing-info/invoices-test.js (hunk 1 above)
Comment (mrsaicharan1) [reply tests/unit/controllers/account/billing-info/invoices-test.js]: Yeah, will remove this test. It got generated automatically when I used the ember CLI to generate the controller.

File: tests/unit/controllers/account/billing-info/invoices-test.js (hunk 2 above)
Comment (mrsaicharan1) [reply tests/unit/controllers/account/billing-info/invoices-test.js]: @Anupam-dagar Some other test seems to be failing :(

File: app/controllers/account/billing-info/invoices.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import Controller from '@ember/controller';
+import { filterBy } from '@ember/object/computed';
+import moment from 'moment';
+import { computed } from '@ember/object';
+
+export default Controller.extend({
+  dueInvoices      : filterBy('model', 'status', 'due'),
+  paidInvoices     : filterBy('model', 'status', 'paid'),
+  upcomingInvoices : computed(function() {
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices.js]: A computed property _must_ not return a promise. This should happen at the route level.

Comment (mrsaicharan1) [reply app/controllers/account/billing-info/invoices.js]: @niranjan94 Moved it to the routes and returned a hash of promises

File: app/routes/account/billing-info/invoices.js [hunk 4]
```diff
@@ -1,4 +1,34 @@
 import Route from '@ember/routing/route';
+import RSVP from 'rsvp';
+import moment from 'moment';
 
 export default Route.extend({
+  async model() {
+    let filterOptions = [];
```
Comment (niranjan94) [inline app/routes/account/billing-info/invoices.js]: this declaration of empty array is unnecessary ..

File: app/routes/account/billing-info/invoices.js [hunk 5]
```diff
@@ -1,4 +1,34 @@
...
+          },
+          {
+            name : 'created-at',
+            op   : 'ge',
+            val  : moment().subtract(30, 'days').toISOString()
+          }
+        ]
+      }
+    ];
+    return RSVP.hash({
```
Comment (niranjan94) [inline app/routes/account/billing-info/invoices.js]: Use async-await instead.

File: app/routes/account/billing-info/invoices.js (hunk 4 above)
Comment (mrsaicharan1) [reply app/routes/account/billing-info/invoices.js]: Travis seems to fail with this message when I remove it.. .`ReferenceError: filterOptions is not defined`

**Comment by mrsaicharan1** [comment]
> Use ember tables?

Yeah, just saw your merged PR. I  considered refactoring after this was merged but I'll do it right now as the ember tables are functional

File: app/controllers/account/billing-info/invoices.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import Controller from '@ember/controller';
+import { filterBy } from '@ember/object/computed';
+
+export default Controller.extend({
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices.js]: Lets start moving to ES6 style classes for all new resources. 
https://github.com/fossasia/open-event-frontend/pull/3169/files

File: app/routes/account/billing-info/invoices.js [hunk 7]
```diff
@@ -1,4 +1,33 @@
 import Route from '@ember/routing/route';
+import RSVP from 'rsvp';
+import moment from 'moment';
 
 export default Route.extend({
```
Comment (niranjan94) [inline app/routes/account/billing-info/invoices.js]: Lets start moving to ES6 style classes for all new resources. 
https://github.com/fossasia/open-event-frontend/pull/3169/files

**Comment by kushthedude** [comment]
Use ember tables for all pending and past too.

**Comment by mrsaicharan1** [comment]
> Use ember tables for all pending and past too.

Pending & past invoices ? What does that mean?

File: app/routes/account/billing-info/invoices.js (hunk 4 above)
Comment (niranjan94) [reply app/routes/account/billing-info/invoices.js]: yes. you should be doing 

```js
const filterOptions = [
      {
        and: [
          {
            name : 'deleted-at',
            op   : 'eq',
            val  : null
          },
          {
            name : 'created-at',
            op   : 'ge',
            val  : moment().subtract(30, 'days').toISOString()
          }
        ]
      }
    ]
```

File: app/templates/account/billing-info/invoices.hbs [hunk 8]
```diff
@@ -1,14 +1,115 @@
...
+                  sortBy=sort_by
+                  sortDir=sort_dir
+                  metaData=model.data.meta
+                  filterOptions=filterOptions
+                  widthConstraint="eq-container"
+                  resizeMode="fluid"
+                  fillMode="equal-column"
+    }}
   </div>
+    {{!-- <table class="ui stackable structured celled compact table">
```
Comment (niranjan94) [inline app/templates/account/billing-info/invoices.hbs]: Remove the commented parts if not required

File: app/controllers/account/billing-info/invoices.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
+import Controller from '@ember/controller';
+import { filterBy } from '@ember/object/computed';
+import { computed } from '@ember/object';
+
+export default class extends Controller {
+  queryParams = ['page', 'per_page'];
+  page = 1;
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices.js]: I think @CosmicCoder96 has added mixins for this boilerplate code. Use that

File: app/controllers/account/billing-info/invoices.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  per_page = 10;
+  search = null;
+  sort_dir = null;
+  sort_by = null;
+  dueInvoices  = filterBy('model.eventInvoices', 'status', 'due');
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices.js]: Should be a decorator

File: app/controllers/account/billing-info/invoices.js [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  paidInvoices = filterBy('model.eventInvoices', 'status', 'paid');
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices.js]: Should be a decorator

**Comment by kushthedude** [comment]
> > Use ember tables for all pending and past too.
> 
> Pending & past invoices ? What does that mean?

You understood it correctly 😄

File: tests/unit/routes/app/templates/account/billing-info/invoices/list-test.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import { module, test } from 'qunit';
+import { setupTest } from 'ember-qunit';
+
+module('Unit | Route | app/templates/account/billing-info/invoices/list', function(hooks) {
+  setupTest(hooks);
+
+  test('it exists', function(assert) {
+    let route = this.owner.lookup('route:app/templates/account/billing-info/invoices/list');
```
Comment (kushthedude) [inline tests/unit/routes/app/templates/account/billing-info/invoices/list-test.js]: It would be great if you can you write render test and Table test as done in others components rather than route test .

File: app/controllers/account/billing-info/invoices.js [hunk 13]
```diff
@@ -0,0 +1,91 @@
+import Controller from '@ember/controller';
+import { filterBy } from '@ember/object/computed';
+import { computed } from '@ember/object';
+import EmberTableMixin from 'open-event-frontend/mixins/ember-table';
+	
+export default class extends Controller.extend(EmberTableMixin) {
+  // queryParams = ['page', 'per_page'];
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices.js]: Remove if not required

File: app/controllers/account/billing-info/invoices.js [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+      {
+        name: 'View Invoice',
+        valuePath: 'invoiceLink'   
+      }
+
+    ];
+  }
+
+  @computed('dueInvoices')
+  get dueInvoiceRows() {
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices.js]: Why is this needed ? Can't `dueInvoices` be used directly ?

File: app/controllers/account/billing-info/invoices.js [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+        dateIssued    : row.createdAt,
+        amount        : row.amount,
+        invoiceLink   : row.invoicePdfUrl
+      });
+    });
+    return rows;
+  }
+
+  @computed('paidInvoices')
+  get paidInvoiceRows() {
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices.js]: Use `paidInvoices` directly

File: app/controllers/account/billing-info/invoices.js [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+        amount        : row.amount,
+        datePaid      : row.completedAt,
+        invoiceLink   : row.invoicePdfUrl
+      });
+    });
+    return rows;
+  }
+
+  @computed('upcomingInvoices')
+  get upcomingInvoiceRows() {
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices.js]: Use `upcomingInvoices` directly

File: app/controllers/account/billing-info/invoices.js (hunk 14 above)
Comment (mrsaicharan1) [reply app/controllers/account/billing-info/invoices.js]: Did you mean that `dueInvoices` can be used directly instead of watching for them?

**Comment by mrsaicharan1** [comment]
@niranjan94 @CosmicCoder96 @uds5501 @shreyanshdwivedi I've integrated ember tables for the event invoice section. This PR is now ready for a review.

File: app/controllers/account/billing-info/invoices.js (hunk 14 above)
Comment (niranjan94) [reply app/controllers/account/billing-info/invoices.js]: Yes.

File: app/controllers/account/billing-info/invoices/list.js [hunk 17]
```diff
@@ -0,0 +1,193 @@
+import Controller from '@ember/controller';
+import { computed } from '@ember/object';
+import EmberTableControllerMixin from 'open-event-frontend/mixins/ember-table-controller';
+
+export default class extends Controller.extend(EmberTableControllerMixin) {
+
+  @computed()
+  get upcomingInvoiceColumns() {
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices/list.js]: These are arrays with plain objects. Can be a normal property instead of a computed property

File: app/controllers/account/billing-info/invoices/list.js [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+      {
+        name      : 'View Invoice',
+        valuePath : 'invoiceLink'
+      }
+
+    ];
+  }
+
+  @computed()
+  get AllColumns() {
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices/list.js]: These are arrays with plain objects. Can be a normal property instead of a computed property

File: app/controllers/account/billing-info/invoices/list.js [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+      {
+        name      : 'Amount',
+        valuePath : 'amount'
+      }
+
+    ];
+  }
+
+  @computed()
+  get paidInvoiceColumns() {
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices/list.js]: These are arrays with plain objects. Can be a normal property instead of a computed property

File: app/controllers/account/billing-info/invoices/list.js [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+      {
+        name      : 'View Invoice',
+        valuePath : 'invoiceLink'
+      }
+
+    ];
+  }
+
+  @computed()
+  get dueInvoiceColumns() {
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices/list.js]: These are arrays with plain objects. Can be a normal property instead of a computed property

File: app/controllers/account/billing-info/invoices/list.js [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+      {
+        name      : 'View Invoice',
+        valuePath : 'invoiceLink'
+      }
+
+    ];
+  }
+
+  @computed()
+  get allInvoiceColumns() {
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices/list.js]: These are arrays with plain objects. Can be a normal property instead of a computed property

File: app/controllers/account/billing-info/invoices/list.js (hunk 17 above)
Comment (mrsaicharan1) [reply app/controllers/account/billing-info/invoices/list.js]: @niranjan94 Replaced these with POJOs.

Comment (mrsaicharan1) [reply app/controllers/account/billing-info/invoices/list.js]: @niranjan94 Replaced these with POJOs. Please have a look.

File: app/controllers/account/billing-info/invoices/list.js [hunk 22]
```diff
@@ -0,0 +1,161 @@
+import Controller from '@ember/controller';
+import { computed } from '@ember/object';
+import EmberTableControllerMixin from 'open-event-frontend/mixins/ember-table-controller';
+
+export default class extends Controller.extend(EmberTableControllerMixin) {
```
Comment (niranjan94) [inline app/controllers/account/billing-info/invoices/list.js]: Check alignment on this file. Seems misaligned.
Also, use the model directly instead of an unnecessary computed property in between

**Comment by shreyanshdwivedi** [comment]
@niranjan94 I'm following changes asked by you on this PR closely and I'll be making the changes to my PR of organizer session table. Will tag you once ready for review again

File: app/controllers/account/billing-info/invoices/list.js [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+  {
+    name      : 'Status',
+    valuePath : 'status'
+  }
+
+];
+
+get dueInvoiceRows() {
+  const rows = [];
+  this.model.data.forEach(row => {
```
Comment (shreyanshdwivedi) [inline app/controllers/account/billing-info/invoices/list.js]: I think we can use a `this.model.data.map` here and for other such iterations too.

**Comment by mrsaicharan1** [comment]
Had to change the route billing-info -> billing to comply with the recent change. Made the requested changes to this. Please have a look. 
Will proceed with the payment workflow as soon as this gets merged.

File: app/controllers/account/billing/invoices/list.js [hunk 24]
```diff
@@ -0,0 +1,156 @@
+import Controller from '@ember/controller';
+import EmberTableControllerMixin from 'open-event-frontend/mixins/ember-table-controller';
+
+export default class extends Controller.extend(EmberTableControllerMixin) {
```
Comment (niranjan94) [inline app/controllers/account/billing/invoices/list.js]: @mrsaicharan1 the spacing is incorrect in this file. Please fix it.

File: app/controllers/account/billing/invoices/list.js [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+      amount      : row.amount,
+      invoiceLink : row.invoicePdfUrl
+    });
+  });
+  return rows;
+}
+
+get paidInvoiceRows() {
+  const rows = [];
+  this.model.data.map(row => {
```
Comment (niranjan94) [inline app/controllers/account/billing/invoices/list.js]: Use model directly

File: app/controllers/account/billing/invoices/list.js [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+      datePaid    : row.completedAt,
+      invoiceLink : row.invoicePdfUrl
+    });
+  });
+  return rows;
+}
+
+get upcomingInvoiceRows() {
+  const rows = [];
+  this.model.data.map(row => {
```
Comment (niranjan94) [inline app/controllers/account/billing/invoices/list.js]: Use model directly

File: app/controllers/account/billing/invoices/list.js [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+      amount      : row.amount,
+      invoiceLink : row.invoicePdfUrl
+    });
+  });
+  return rows;
+}
+
+get allInvoiceRows() {
+  const rows = [];
+  this.model.data.map(row => {
```
Comment (niranjan94) [inline app/controllers/account/billing/invoices/list.js]: Use model directly

File: app/controllers/account/billing/invoices/list.js (hunk 25 above)
Comment (mrsaicharan1) [reply app/controllers/account/billing/invoices/list.js]: Do you mean `this.model.data.map(row => {...`? It returns undefined when I do that.

**Comment by mrsaicharan1** [comment]
@CosmicCoder96 Please review this. Would want to proceed with the other parts of this asap

File: app/controllers/account/billing/invoices/list.js (hunk 25 above)
Comment (CosmicCoder96) [reply app/controllers/account/billing/invoices/list.js]: No, check the other PRs.

File: app/controllers/account/billing/invoices/list.js [hunk 28]
```diff
@@ -0,0 +1,155 @@
...
+    name      : 'Amount Due',
+    valuePath : 'amount'
+  },
+  {
+    name      : 'View Invoice',
+    valuePath : 'invoiceLink'
+  }
+
+];
+allInvoiceColumns =  [
```
Comment (CosmicCoder96) [inline app/controllers/account/billing/invoices/list.js]: make them computed properties.

Comment (mrsaicharan1) [reply app/controllers/account/billing/invoices/list.js]: @CosmicCoder96 @niranjan94 Had mentioned that computed properties won't be required here. POJOs are enough. [In one of his reviews]

**Comment by mrsaicharan1** [comment]
@CosmicCoder96 I could've used the model directly by passing in `rows=model.data` if the columns were similar for 3 of the categories. But here, for due, upcoming & paid invoices, the row entries which are being pushed are different.

File: app/models/event-invoice.js [hunk 29]
```diff
@@ -26,6 +26,8 @@ export default ModelBase.extend({
   expYear      	: attr('number'),
   amount        : attr('number'),
   completedAt   : attr('moment'),
+  invoicePdfUrl : attr('string'),
+  eventId       : attr('number'),
```
Comment (uds5501) [inline app/models/event-invoice.js]: Do check if it can be achieved using `event.get('originalId')` or something similar

**Comment by mrsaicharan1** [comment]
Yeah this is not required anyways ... I'll remove it

**Comment by mrsaicharan1** [comment]
@uds5501 @shreyanshdwivedi @CosmicCoder96 I was able to obtain the event logo & name by creating a cell component & passing in the record. Is this the right way?

File: app/models/event-invoice.js (hunk 29 above)
Comment (CosmicCoder96) [reply app/models/event-invoice.js]: this is the actual event invoice model, either the server is returning the eventId or not, if it is, then it's definition here is not optional.

File: app/routes/account/billing/invoices/list.js [hunk 30]
```diff
@@ -0,0 +1,73 @@
...
+              val  : moment().subtract(30, 'days').toISOString()
+            }
+          ]
+        }
+      ];
+    }
+
+    filterOptions = this.applySearchFilters(filterOptions, params, searchField);
+
+    let queryObject = {
```
Comment (CosmicCoder96) [inline app/routes/account/billing/invoices/list.js]: don't break the pattern, every other table has the name `queryString'

File: app/routes/account/billing/invoices/list.js [hunk 31]
```diff
@@ -0,0 +1,73 @@
...
+      include        : 'event',
+      filter         : filterOptions,
+      'page[size]'   : params.per_page || 10,
+      'page[number]' : params.page || 4
```
Comment (CosmicCoder96) [inline app/routes/account/billing/invoices/list.js]: Why would you start with page 4?

**Comment by CosmicCoder96** [comment]
> I was able to obtain the event logo & name by creating a cell componen

Does it actually work, did you test it?

**Comment by mrsaicharan1** [comment]
> > I was able to obtain the event logo & name by creating a cell componen
> 
> Does it actually work, did you test it?

Yes, I've updated the screenshot too. Please have a look. Also, the `eventId` field was not required & therefore I removed it.

**Comment by mrsaicharan1** [comment]
@pradeepgangwar @uds5501 @shreyanshdwivedi This is ready for another review.

File: app/templates/account/billing/invoices/list.hbs [hunk 32]
```diff
@@ -0,0 +1,65 @@
+{{#if (eq model.params.invoice_status 'due') }}
```
Comment (CosmicCoder96) [inline app/templates/account/billing/invoices/list.hbs]: Too much repitition.

**Comment by mrsaicharan1** [comment]
I have shifted the logic to controllers & removed repetitive template code. Please have a look now.

**Comment by mrsaicharan1** [comment]
@CosmicCoder96 Made the requested changes. Please have a look.
